src/test/java/uk/gov/pay/api/validation/PaymentRequestValidatorTest.java

```diff
@@ -75,47 +82,160 @@ public void validateMaximumAmount_shouldSuccessValue() {
+
+        expectedException.expect(PaymentValidationException.class);
+        expectedException.expect(aValidationExceptionContaining("P0102", "Invalid attribute value: line2. Must be less than or equal to 255 characters length"));
+
+        paymentRequestValidator.validate(createPaymentRequest);
+    }
+
+    @Test
+    public void validatePostCodeMaxLength_shouldFailValue() {
+        String invalidMaxLengthEmail = randomAlphanumeric(PaymentRequestValidator.POSTCODE_MAX_LENGTH+ 1);
+        CreatePaymentRequest createPaymentRequest = createPaymentRequestBuilderWithAgreementId().postcode(invalidMaxLengthEmail).build();
+
+        expectedException.expect(PaymentValidationException.class);
+        expectedException.expect(aValidationExceptionContaining("P0102", "Invalid attribute value: postcode. Must be less than or equal to 25 characters length"));
+
+        paymentRequestValidator.validate(createPaymentRequest);
+    }
+
+    @Test
+    public void validateCityMaxLength_shouldFailValue() {
+        String invalidMaxLengthEmail = randomAlphanumeric(PaymentRequestValidator.CITY_MAX_LENGTH+ 1);
+        CreatePaymentRequest createPaymentRequest = createPaymentRequestBuilderWithAgreementId().city(invalidMaxLengthEmail).build();
+
         expectedException.expect(PaymentValidationException.class);
+        expectedException.expect(aValidationExceptionContaining("P0102", "Invalid attribute value: city. Must be less than or equal to 255 characters length"));
+
+        paymentRequestValidator.validate(createPaymentRequest);
+    }
+
+    @Test
+    public void validateCountryMaxLength_shouldFailValue() {
+        String invalidMaxLengthEmail = randomAlphanumeric(PaymentRequestValidator.COUNTRY_EXACT_LENGTH+ 1);
+        CreatePaymentRequest createPaymentRequest = createPaymentRequestBuilderWithAgreementId().country(invalidMaxLengthEmail).build();
+
+        expectedException.expect(PaymentValidationException.class);
+        expectedException.expect(aValidationExceptionContaining("P0102", "Invalid attribute value: country. Must be exactly 2 characters length"));
+
+        paymentRequestValidator.validate(createPaymentRequest);
+    }
+
+    @Test
+    public void validateCountryMinLength_shouldFailValue() {
+        String invalidMaxLengthEmail = randomAlphanumeric(1);
+        CreatePaymentRequest createPaymentRequest = createPaymentRequestBuilderWithAgreementId().country(invalidMaxLengthEmail).build();
+
+        expectedException.expect(PaymentValidationException.class);
+        expectedException.expect(aValidationExceptionContaining("P0102", "Invalid attribute value: country. Must be exactly 2 characters length"));
+
+        paymentRequestValidator.validate(createPaymentRequest);
+    }
+    
+    @Test
+    public void validateCountryEmpty_shouldPass() {
+        CreatePaymentRequest createPaymentRequest = createPaymentRequestBuilderWithAgreementId().country("").build();
+
```

Comment:
It’s great that we’re reducing duplication of tests that test basically the same validation logic at different levels but it seems the functionality covered by the old `deserialize_shouldThrowValidationException_whenReturnUrlIsAMalformedUrl` test has been lost in the migration.

Reply:
Great catch, thank you reviewing so diligently its much appreciated! I've added this in.